onnxruntime/contrib_ops/cuda/quantization/dequantize_blockwise.cu

```diff
@@ -18,112 +18,216 @@
+ * @tparam qbits          number of bits in each quantized element
+ * @tparam Columnwise     true:  elements in a block come from one single column
+ *                        false: elements in a block come from one single row
+ */
+template <
+  typename ElementT,
+  int32_t block_size,
+  int32_t qbits,
+  bool Columnwise>
+struct BlkQuantTraits {
+  // number of qbit elements to pack into whole bytes
+  static constexpr int kPackSize = (qbits == 8) ? 1 : (qbits == 4) ? 2 : (qbits == 2) ? 4 : 0;
+  static_assert(kPackSize != 0, "Packing to whole bytes not supported for this qbits!");
+
+  using QuantBlk = std::conditional_t<Columnwise, Shape2D<block_size, 1>, Shape2D<1, block_size>>;
+  using ThreadBlk = Shape2D<QuantBlk::kRow * kPackSize, QuantBlk::kColumn>;
+};
+
+template <
+  typename ElementT,
+  int32_t block_size,
+  int32_t qbits,
+  bool Columnwise>
+__global__
+void dequantizeThread(ElementT* dst, const uint8_t* weights, const ElementT* scales,
+  const uint8_t* zero_points, int rows,  int columns,  int thrd_col_blks) {
+
+  using QuantBlk = typename BlkQuantTraits<ElementT, block_size, qbits, Columnwise>::QuantBlk;
+  using ThreadBlk = typename BlkQuantTraits<ElementT, block_size, qbits, Columnwise>::ThreadBlk;
+
+  const auto block_idx = blockIdx.x * blockDim.x + threadIdx.x;
+  const auto row_blks = (rows + QuantBlk::kRow - 1) / QuantBlk::kRow;
+
+  const auto meta_rows = (rows + QuantBlk::kRow - 1) / QuantBlk::kRow;
+  const auto meta_cols = (columns + QuantBlk::kColumn - 1) / QuantBlk::kColumn;
+
+  // quantized matrix is stored in column major, packed by column
+  const auto q_rows = (meta_rows * QuantBlk::kRow * qbits + 7) / 8;
+  const auto q_cols = meta_cols * QuantBlk::kColumn;
+
+  int32_t r_blk_idx = static_cast<int32_t>(block_idx / thrd_col_blks);
+  int32_t c_blk_idx = static_cast<int32_t>(block_idx % thrd_col_blks);
+
+  int32_t r = r_blk_idx * ThreadBlk::kRow;
+  int32_t c = c_blk_idx * ThreadBlk::kColumn;
+
+  int32_t r_end = std::min(r + ThreadBlk::kRow, rows);
+  int32_t c_end = std::min(c + ThreadBlk::kColumn, columns);
+
+  for (int32_t j = c; j < c_end; ++j) {
+    const int32_t meta_col = j / QuantBlk::kColumn;
+
+    // !! 4b specific code
+    // the whole loop is 4b specific due to sub 8 bit packing
+    // and unpacking. We can potentially make this qbits generic
+    // by wraping the packing/unpacking code like cutlass::Array
+    static_assert(qbits == 4, "Only 4b block quantization is supported!");
+
+    for (int32_t i = r; i < r_end; i += 2) {
+      const int32_t meta_row = i / QuantBlk::kRow;
+
+      const float scale0 = static_cast<float>(scales[meta_col * row_blks + meta_row]);
+      const int zp_pair = (zero_points == nullptr)
+            ? 0x88
+            : zero_points[meta_col * ((row_blks + 1) / 2) + meta_row / 2];
+      const int zp0 = (meta_row & 1) ? (zp_pair >> 4) : (zp_pair & 0xf);
+
+      const uint8_t vi0 = weights[j * q_rows + i / 2] & 0xf;
```

Comment:
did you benchmark? strided memory access is inefficient in gpu.

Reply:
inner loop index is 2, this is sequential access not strided?

Reply:
is the inner loop size is ThreadBlk::kRow/2, which is block_size for columnsize?

Reply:
for column-wise, the access pattern is exactly the same as your dequant

Reply:
new changes split the implementation, column-wise still using old implementation